src/box/lua/net_box.c

```diff
@@ -1234,10 +1256,8 @@ netbox_transport_communicate(struct netbox_transport *transport, size_t limit)
 			box_error_raise(ER_NO_CONNECTION, "Error injection");
 			return -1;
 		});
-		if (fiber_is_cancelled()) {
-			diag_set(FiberIsCancelled);
+		if (netbox_worker_check_after_yield(transport) != 0)
```

Comment:
> The worker fiber can yield in a number of places: after `coio_wait` and during Lua calls.
> 
> During the worker's yield, the connection can get closed. When the connection is closed, its receive buffer is reset. If there was some pending response that was partially retrieved (e.g., a large select), then after resetting the buffer we will read some inconsistent data. We must not allow this to happen, so let's check for this case after returning from places where the worker can yield.

`luaT_netbox_transport_stop` cancels the worker fiber right after calling `netbox_transport_set_error`, which resets the buffer, so the worker fiber should exit immediately after `coio_wait`, without trying to read the buffer any further. Please describe what happens exactly. Maybe, it'll help preparing a test.

Reply:
> `luaT_netbox_transport_stop` cancels the worker fiber right after calling `netbox_transport_set_error`, which resets the buffer, so the worker fiber should exit immediately after `coio_wait`.

This is not always true. In between `netbox_transport_set_error` and `fiber_cancel`, `netbox_transport_on_change` is called:
https://github.com/tarantool/tarantool/blob/3daf2399dd5b2b1987e258918376ff9be87e9b01/src/box/lua/net_box.c#L3188-L3194

`netbox_transport_on_change`, in turn, can call a user-provided `on_disconnect` trigger, which can yield.

So there is a chance that the network fiber returns from `coio_wait` and does not get cancelled, while the connection is already closed.

> Please describe what happens exactly.

Unlike the bug from the second patch, I was not able to materialize an inconsistent state that would crash the network fiber. However, it is fairly easy to observe that the scenario I describe above is possible (I can bring a snippet with `print`s, if you like).

Reply:
> `netbox_transport_on_change`, in turn, can call a user-provided `on_disconnect` trigger, which can yield.
> 
> So there is a chance that the worker fiber returns from `coio_wait` and does not get cancelled, while the connection is already closed.

Then I think that the issue should be possible to reproduce and test properly as follows:
1. Create a fake Tarantool server that sends only a part of a fixheader (half `MP_UINT`) in reply to some special request, then stops until further notice. Maybe, it can even be done with a real Tarantool server using `box.iproto.override`.
2. Establish a net.box connection to this server and send this special request.
3. Install `on_disconnect` trigger that sleeps.
4. Close the connection, then notify the Tarantool server to send the rest of the packet, then resume the `on_disconnect` trigger.

> > worker fiber should exit immediately after `coio_wait`, without trying to read the buffer any further.
> 
> AFAIC, the worker fiber does not always get blocked at `coio_wait`. If there is data ready in the socket, it can start reading it, let's say reading out `request1` and part of `request2`. Later, it can yield from a Lua call, return from it, not being aware that the connection got closed and that the receive buffer got reset, and continue reading from the socket (assuming another part of `request2` data arrives by this moment), again not being blocked by `coio_wait`. Finally, it will start reading the receive buffer, which contains the middle of `request2` (not the beginning, since it was already read out and the buffer got reset), and crash.

Can you fix this issue by cancelling and joining the worker fiber before resetting the buffer?

Reply:
> Then I think that the issue should be possible to reproduce and test properly as follows.

Thanks for the suggestion! I will try it out and return to you.

> Can you fix this issue by cancelling and joining the worker fiber before resetting the buffer?

AFAIC, this will not work if we close the connection from inside the worker fiber (either do to a trigger, or garbage collection).

Reply:
> > Can you fix this issue by cancelling and joining the worker fiber before resetting the buffer?
> 
> AFAIC, this will not work if we close the connection from inside the worker fiber (either due to a trigger, or due to garbage collection).

If we do that, we should raise a Lua exception that would force the worker fiber to exit by itself.

Reply:
> If we do that, we should raise a Lua exception that would force the worker fiber to exit by itself.

But then again, that would mean we would interfere with the `on_disconnect` trigger (AFAIC, we would not even run it with the solution you suggest).

Reply:
Here is a reproducer for crash after returning from `coio_wait`:

```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

fiber = require('fiber')
nb = require('net.box')
socket = require('socket')

local cond = fiber.cond()
local handler = function(fd)
    -- Greeting.
    fd:write(box.iproto.encode_greeting())
    -- Identification.
    fd:write(box.iproto.encode_packet({request_type =
                                       bit.bor(box.iproto.type.TYPE_ERROR,
                                               box.error.UNKNOWN_REQUEST_TYPE)}))
    -- ce 00 00 00 03 81 00 00
    local packet = box.iproto.encode_packet({request_type = box.iproto.type.OK})
    -- ce 00 00 00 03
    local partial_packet = packet:sub(1, 5)
    -- 81 00 00 ce 00 00 00 03 81 00 00
    local leftover_packet = packet:sub(6) .. packet
    fd:write(partial_packet)
    cond:wait()
    fd:write(leftover_packet)
end

srv = socket.tcp_server('localhost', 3301, handler)

c = nb.new(srv:name().port, {fetch_schema = false})

c:on_disconnect(function()
    fiber.sleep(600)
end)

fiber.create(function()
    c:close()
end)

cond:signal()

fiber.sleep(600)

os.exit()
```

Here is a reproducer for crash after returning from `on_connect` trigger:
```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

fiber = require('fiber')
nb = require('net.box')
socket = require('socket')

local cond1 = fiber.cond()
local cond2 = fiber.cond()
local cond3 = fiber.cond()

local handler = function(fd)
    -- Greeting.
    fd:write(box.iproto.encode_greeting())
    -- Identification.
    fd:read(1)
    -- Identification.
    fd:write(box.iproto.encode_packet({request_type =
                                       bit.bor(box.iproto.type.TYPE_ERROR,
                                               box.error.UNKNOWN_REQUEST_TYPE)}))
    -- ce 00 00 00 03 81 00 00
    local packet = box.iproto.encode_packet({request_type = box.iproto.type.OK})
    -- ce 00 00 00 03
    local partial_packet = packet:sub(1, 5)
    -- 81 00 00 ce 00 00 00 03 81 00 00
    local leftover_packet = packet:sub(6) .. packet
    fd:write(partial_packet)
    cond1:wait()
    fd:write(leftover_packet)
end

srv = socket.tcp_server('localhost', 3301, handler)

c = nb.new(srv:name().port, {wait_connected = false, fetch_schema = false})
c:on_connect(function()
    cond2:signal()
    cond3:wait()
end)

c:on_disconnect(function()
    fiber.sleep(600)
end)

cond2:wait()

fiber.create(function()
    c:close()
end)

cond1:signal()
cond3:signal()

fiber.sleep(600)

os.exit()
```

Reply:
Then let's try to call triggers from the fiber. I mean something like the patch below.

<details><summary>patch</summary>

```diff
diff --git a/src/box/lua/net_box.c b/src/box/lua/net_box.c
index 1f1c5f7b65a6..c476c9c8df6f 100644
--- a/src/box/lua/net_box.c
+++ b/src/box/lua/net_box.c
@@ -3082,7 +3082,7 @@ netbox_worker_f(va_list ap)
 	assert(transport->self_ref != LUA_NOREF);
 	const double reconnect_after = !uri_is_nil(&transport->opts.uri) ?
 				       transport->opts.reconnect_after : 0;
-	while (!fiber_is_cancelled()) {
+	while (true) {
 		if (netbox_transport_connect(transport) == 0) {
 			int rc = luaT_cpcall(L, netbox_connection_handler_f,
 					     transport);
@@ -3091,8 +3091,13 @@ netbox_worker_f(va_list ap)
 			(void)rc;
 			iostream_close(&transport->io);
 		}
-		if (transport->state == NETBOX_CLOSED)
+		if (fiber_is_cancelled()) {
+			box_error_raise(ER_NO_CONNECTION, "Connection closed");
+			netbox_transport_set_error(transport);
+			transport->state = NETBOX_CLOSED;
+			netbox_transport_on_state_change_pcall(transport, L);
 			break;
+		}
 		netbox_transport_set_error(transport);
 		transport->state = (reconnect_after > 0 ?
 				    NETBOX_ERROR_RECONNECT : NETBOX_ERROR);
@@ -3179,17 +3184,6 @@ luaT_netbox_transport_stop(struct lua_State *L)
 			fiber_cond_wait(&transport->on_send_buf_empty);
 		transport->is_closing = false;
 	}
-	/*
-	 * While we were waiting for the send buffer to be empty,
-	 * the state could change.
-	 */
-	if (transport->state != NETBOX_CLOSED &&
-	    transport->state != NETBOX_ERROR) {
-		box_error_raise(ER_NO_CONNECTION, "Connection closed");
-		netbox_transport_set_error(transport);
-		transport->state = NETBOX_CLOSED;
-		netbox_transport_on_state_change(transport, L);
-	}
 	/* Cancel the worker fiber. */
 	if (transport->worker != NULL) {
 		fiber_cancel(transport->worker);
```

</details>

It should fix the crashes. If running `on_disconnect` triggers from a cancelled fiber isn't okay, we could use some other way to signal the worker fiber to stop.

Reply:
This solution is bad:
1. **With this solution**, the following snippet hangs:
```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local net = require('net.box')

box.cfg{listen = 3301}
box.schema.user.grant('guest', 'execute', 'universe')

local c = net.connect(3301, {wait_connected = false})
c:close()
c:ping()

os.exit()
```
2. **With this solution**, the following snippet crashes:
``lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local fiber = require('fiber')
local net = require('net.box')

box.cfg{listen = 3301}
box.schema.user.grant('guest', 'execute', 'universe')

local c = net.connect(3301, {wait_connected = true})
c:on_disconnect(function()
    fiber.yield()
end)

c:close()

os.exit()
```

Reply:
I think problem 2 could be solved by waiting for the worker fiber to exit after cancelling it in `luaT_netbox_transport_stop`.

Reply:
AFAIC, that would cause a deadlock in case `transport_stop` gets called from the worker fiber (e.g., in some trigger).

Reply:
`luaT_netbox_transport_stop` raises the `FiberIsCancelled` error if called from the worker fiber. It shouldn't wait in this case, apparently.

Reply:
Problem 2 still remains in this case, since the user can still observe the connection state as active from other fibers (let's say, we have a yield after a close in the trigger).

Reply:
I don't think this is a problem: `close` will raise an error in this case so there's no guarantee that the connection is closed immediately.

Reply:
`close` will not raise after #9838 is merged, because raising errors from `luaT_netbox_transport_stop` is UB, see #9678.

But we could patch the Lua part of `close` to check the connection status and throw an error:
https://github.com/tarantool/tarantool/blob/58f5ffe43bffa0901d8d577bbb4d523c4e7e682a/src/box/lua/net_box.lua#L838-L841

Reply:
`close` should raise an exception only if it's called from the worker fiber - in order to cancel itself. This code can't be called from gc AFAICS because the worker fiber keeps a reference to the connection, i.e. we just need to update the condition in `close` a little:

```lua
diff --git a/src/box/lua/net_box.c b/src/box/lua/net_box.c
index 8b0f693e2806..38ae601201c4 100644
--- a/src/box/lua/net_box.c
+++ b/src/box/lua/net_box.c
@@ -3206,7 +3206,8 @@ luaT_netbox_transport_stop(struct lua_State *L)
 	if (transport->worker != NULL) {
 		fiber_cancel(transport->worker);
 		/* Check if we cancelled ourselves. */
-		luaL_testcancel(L);
+		if (transport->worker == fiber())
+			luaL_testcancel(L);
 	}
 	return 0;
 }
```

Reply:
This code can't be called from gc AFAICS because the worker fiber keeps a reference to the connection, i.e. we just need to update the condition in `close` a little.

No, this code can definitely be called from garbage collection — that's what happens in #9621. To be honest, I do not understand how the referencing works here, this snippet is especially obscure:
https://github.com/tarantool/tarantool/blob/58f5ffe43bffa0901d8d577bbb4d523c4e7e682a/src/box/lua/net_box.lua#L419-L439

And this snippet seems to not even work properly, since #9782 exists.

As you already pointed out, the exception thrown from `stop` will be caught by `pcall` in the `gc` hook. Since the connection is garbage collected, nobody can observe its state, so I think it's totally fine that we have it in this intermediate `closing` state for some time.

Reply:
> I suggest we go for the patch you suggested, which moves the trigger call after the end of the worker loop, and joins the worker fiber, if it isn't the current fiber. Otherwise we throw an exception to indicate that close did not work properly. The user can either let it fly and stop the worker loop, or catch it at his own peril.

Sounds good to me.

Reply:
Updated the PR as per our discussion. Note that it didn't help fix #9785.